MqttIOTest.testRead is flaky (sickbayed)

Failing job url:

[https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/1239/testReport/org.apache.beam.sdk.io.mqtt/MqttIOTest/testRead/](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/1239/testReport/org.apache.beam.sdk.io.mqtt/MqttIOTest/testRead/)


 Job history url:

https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/1234/testReport/junit/org.apache.beam.sdk.io.mqtt/MqttIOTest/testRead/history/


 Relevant log:

java.lang.AssertionError: MqttIO.Read/Read(UnboundedMqttSource)/StripIds/ParMultiDo(StripIds).output: Expected: iterable over [[<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <48\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <49\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <50\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <51\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <52\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <53\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <54\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <55\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <56\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <57\>]] in any order but: No item matches: [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <48\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <49\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <50\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <51\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <52\>] in [[<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <53\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <54\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <55\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <56\>], [<84\>, <104\>, <105\>, <115\>, <32\>, <105\>, <115\>, <32\>, <116\>, <101\>, <115\>, <116\>, <32\>, <57\>]] at org.apache.beam.sdk.testing.PAssert$PAssertionSite.capture(PAssert.java:168) at org.apache.beam.sdk.testing.PAssert.that(PAssert.java:374) at org.apache.beam.sdk.testing.PAssert.that(PAssert.java:366) at org.apache.beam.sdk.io.mqtt.MqttIOTest.testRead(MqttIOTest.java:148) at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.lang.reflect.Method.invoke(Method.java:498) at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50) at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47) at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298) at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292) at java.util.concurrent.FutureTask.run(FutureTask.java:266)

Imported from Jira [BEAM-5150](https://issues.apache.org/jira/browse/BEAM-5150). Original Jira may contain additional context.
Reported by: Ardagan.

--- a/sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java
+++ b/sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java
@@ -30,6 +30,7 @@ import java.util.NoSuchElementException;
 import java.util.Objects;
 import java.util.UUID;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
 import org.apache.beam.sdk.coders.ByteArrayCoder;
 import org.apache.beam.sdk.coders.Coder;
 import org.apache.beam.sdk.coders.SerializableCoder;
@@ -45,6 +46,7 @@ import org.apache.beam.sdk.values.PDone;
 import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.annotations.VisibleForTesting;
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.fusesource.mqtt.client.BlockingConnection;
+import org.fusesource.mqtt.client.FutureConnection;
 import org.fusesource.mqtt.client.MQTT;
 import org.fusesource.mqtt.client.Message;
 import org.fusesource.mqtt.client.QoS;
@@ -431,8 +433,7 @@ public class MqttIO {
         client = spec.connectionConfiguration().createClient();
         LOG.debug("Reader client ID is {}", client.getClientId());
         checkpointMark.clientId = client.getClientId().toString();
-        connection = client.blockingConnection();
-        connection.connect();
+        connection = createConnection(client);
         connection.subscribe(
             new Topic[] {new Topic(spec.connectionConfiguration().getTopic(), QoS.AT_LEAST_ONCE)});
         return advance();
@@ -569,8 +570,7 @@ public class MqttIO {
         LOG.debug("Starting MQTT writer");
         client = spec.connectionConfiguration().createClient();
         LOG.debug("MQTT writer client ID is {}", client.getClientId());
-        connection = client.blockingConnection();
-        connection.connect();
+        connection = createConnection(client);
       }
 
       @ProcessElement
@@ -589,5 +589,21 @@ public class MqttIO {
         }
       }
     }
+  }
+
+  /** Create a connected MQTT BlockingConnection from given client, aware of connection timeout. */
+  static BlockingConnection createConnection(MQTT client) throws Exception {
+    FutureConnection futureConnection = client.futureConnection();
+    org.fusesource.mqtt.client.Future<Void> connecting = futureConnection.connect();
+    while (true) {
+      try {
+        connecting.await(1, TimeUnit.MINUTES);
+      } catch (TimeoutException e) {
+        LOG.warn("Connection to MQTT broker pending after waiting for 1 minute");
+        continue;
+      }
+      break;
+    }
+    return new BlockingConnection(futureConnection);
   }
 }

--- a/sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java
+++ b/sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java
@@ -599,7 +599,7 @@ public class MqttIO {
       try {
         connecting.await(1, TimeUnit.MINUTES);
       } catch (TimeoutException e) {
-        LOG.warn("Connection to MQTT broker pending after waiting for 1 minute");
+        LOG.warn("Connection to {} pending after waiting for 1 minute", client.getHost());
         continue;
       }
       break;

--- a/sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java
+++ b/sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java
@@ -142,8 +142,7 @@ public class MqttIOTest {
     publisherThread.join();
   }
 
-  @Test(timeout = 30 * 1000)
-  @Ignore("https://github.com/apache/beam/issues/19092 Flake Non-deterministic output.")
+  @Test(timeout = 40 * 1000)
   public void testRead() throws Exception {
     PCollection<byte[]> output =
         pipeline.apply(
@@ -180,12 +179,12 @@ public class MqttIOTest {
                         + "messages ...");
                 boolean pipelineConnected = false;
                 while (!pipelineConnected) {
-                  Thread.sleep(1000);
                   for (Connection connection : brokerService.getBroker().getClients()) {
                     if (connection.getConnectionId().startsWith("READ_PIPELINE")) {
                       pipelineConnected = true;
                     }
                   }
+                  Thread.sleep(2000);
                 }
                 for (int i = 0; i < 10; i++) {
                   publishConnection.publish(

--- a/sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java
+++ b/sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java
@@ -150,7 +150,7 @@ public class MqttIOTest {
                 .withConnectionConfiguration(
                     MqttIO.ConnectionConfiguration.create("tcp://localhost:" + port, "READ_TOPIC")
                         .withClientId("READ_PIPELINE"))
-                .withMaxReadTime(Duration.standardSeconds(3)));
+                .withMaxReadTime(Duration.standardSeconds(5)));
     PAssert.that(output)
         .containsInAnyOrder(
             "This is test 0".getBytes(StandardCharsets.UTF_8),
@@ -184,7 +184,7 @@ public class MqttIOTest {
                       pipelineConnected = true;
                     }
                   }
-                  Thread.sleep(2000);
+                  Thread.sleep(1000);
                 }
                 for (int i = 0; i < 10; i++) {
                   publishConnection.publish(

